runtime/ISO_Fortran_binding.cc

```diff
@@ -193,20 +194,149 @@ int CFI_is_contiguous(const CFI_cdesc_t *descriptor) {
   return 1;
 }
 
+static inline bool IsAssumedSize(const CFI_cdesc_t *dv) {
+  return dv->rank > 0 && dv->dim[dv->rank - 1].extent == -1;
+}
+
 int CFI_section(CFI_cdesc_t *result, const CFI_cdesc_t *source,
     const CFI_index_t lower_bounds[], const CFI_index_t upper_bounds[],
     const CFI_index_t strides[]) {
-  return CFI_INVALID_DESCRIPTOR;  // TODO
+  CFI_index_t extent[CFI_MAX_RANK];
+  CFI_index_t actualStride[CFI_MAX_RANK];
+  CFI_rank_t resRank{0};
+  char *shiftedBaseAddr{static_cast<char *>(source->base_addr)};
+
+  if (source->rank == 0) {
+    return CFI_INVALID_RANK;
+  }
+  if (IsAssumedSize(source) && upper_bounds == nullptr) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if ((result->type != source->type) ||
+      (result->elem_len != source->elem_len)) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if (result->attribute == CFI_attribute_allocatable) {
+    return CFI_INVALID_ATTRIBUTE;
+  }
+  if (source->base_addr == nullptr) {
+    return CFI_ERROR_BASE_ADDR_NULL;
+  }
+
+  bool isZeroSized{false};
+  for (int j{0}; j < source->rank; ++j) {
+    const CFI_dim_t &dim = source->dim[j];
+    const CFI_index_t srcLB = dim.lower_bound;
+    const CFI_index_t srcUB = srcLB + dim.extent - 1;
+    const CFI_index_t lb = (lower_bounds != nullptr) ? lower_bounds[j] : srcLB;
+    const CFI_index_t ub = (upper_bounds != nullptr) ? upper_bounds[j] : srcUB;
+    const CFI_index_t stride = (strides != nullptr) ? strides[j] : 1;
```

Comment:
I believe "var{expr}" initializer syntax is preferred to "var = expr"

Reply:
Thanks, changed as recommended

Reply:
No other reason than readability, but it this to be a personal opinion. I will remove them everywhere to be consistent.

Reply:
The needless parentheses are not a big problem.  I think it is best to avoid adding unnecessary details like those parentheses because they can be distracting to the reader of the code, who is going to wonder why they are there since the expression is sufficiently clear without them.

But I do use parentheses that aren't strictly necessary for a correct parse of an expression when there are operations being used that are commonly misunderstood by many C programmers.  For example, I will write `x = y << (z+1);` even though addition binds more tightly than shifts do, simply to avoid a case that has led to bugs in the past.

GCC will emit warnings when parentheses are not used on `(a && b) || c` though, and I really don't like having to use them there, and do so just to get a clean build.